File: algorithms/golang/438.找到字符串中所有字母异位词.go
/*
 * @lc app=leetcode.cn id=438 lang=golang
 *
 * [438] 找到字符串中所有字母异位词
 */

// @lc code=start
func findAnagrams(s string, p string) []int {
	ret := []int{}
	s_count, p_count, s_len, p_len := [26]int{}, [26]int{}, len(s), len(p)
	if s_len < p_len {
		return ret
	}
	for i, ch := range p {
		p_count[ch-'a']++
		s_count[s[i]-'a']++
	}

	if s_count == p_count {
		ret = append(ret, 0)
	}
	for i := 1; i < len(s)-p_len+1; i++ {
		s_count[s[i-1]-'a']--
		s_count[s[i+p_len-1]-'a']++
		if s_count == p_count {
			ret = append(ret, i)
		}
	}
	return ret
}

// @lc code=end

File: algorithms/js/438.找到字符串中所有字母异位词.js
/*
 * @lc app=leetcode.cn id=438 lang=javascript
 *
 * [438] 找到字符串中所有字母异位词
 */

// @lc code=start
/**
 * @param {string} s
 * @param {string} p
 * @return {number[]}
 */
var findAnagrams = function(s, p) {
  const sLen = s.length,
    pLen = p.length;
  const ret = [];
  if (sLen < pLen) {
    return ret;
  }

  const count = Array.from({ length: 26 }, () => 0),
    tempCount = Array.from({ length: 26 }, () => 0);
  for (let i = 0; i < p.length; i++) {
    count[p.charCodeAt(i) - 97]++;
    tempCount[s.charCodeAt(i) - 97]++;
  }
  if (isEqualArray(count, tempCount)) {
    ret.push(0);
  }
  let start = 1;
  while (start <= sLen - pLen) {
    tempCount[s.charCodeAt(start - 1) - 97]--;
    tempCount[s.charCodeAt(start + pLen - 1) - 97]++;
    if (isEqualArray(count, tempCount)) {
      ret.push(start);
    }
    start++;
  }
  return ret;
};
function isEqualArray(arr1, arr2) {
  if (arr1.length !== arr2.length) {
    return false;
  }
  for (let i = 0; i < arr1.length; i++) {
    if (arr1[i] !== arr2[i]) {
      return false;
    }
  }
  return true;
}
// @lc code=end
